sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala

```diff
@@ -459,11 +459,28 @@ object TableOutputResolver extends SQLConfHelper with Logging {
     }
 
     if (resKey.length == 1 && resValue.length == 1) {
-      val keyFunc = LambdaFunction(resKey.head, Seq(keyParam))
-      val valueFunc = LambdaFunction(resValue.head, Seq(valueParam))
-      val newKeys = ArrayTransform(MapKeys(nullCheckedInput), keyFunc)
-      val newValues = ArrayTransform(MapValues(nullCheckedInput), valueFunc)
-      Some(Alias(MapFromArrays(newKeys, newValues), expected.name)())
+      // If the key and value expressions have not changed, we just check original map field.
+      // Otherwise, we construct a new map by adding transformations to the keys and values.
+      if (resKey.head == keyParam && resValue.head == valueParam) {
+        Some(
+          Alias(nullCheckedInput, expected.name)(
+            nonInheritableMetadataKeys =
+              Seq(CharVarcharUtils.CHAR_VARCHAR_TYPE_STRING_METADATA_KEY)))
```

Comment:
Could you explain why you want remove the `CHAR_VARCHAR_TYPE_STRING_METADATA_KEY` from metadata?

Reply:
I just want to be consistent with #47843

Reply:
@viirya Could you help me understand why remove the `CHAR_VARCHAR_TYPE_STRING_METADATA_KEY` from metadata?

Reply:
Hmm, I don't remember the exact reason on the removal. As it is the input expression used to write to the table, the metadata is useless (because the writer would check on table columns).